### ISSUE TITLE
High-order mappings with manifolds are broken

### ISSUE THREAD

**Issue by agrayver** [dealii/dealii#1646]
I vote to close #338 and refer to this issue as there is now more details and new test program. 

I have created a program which projects a simple polynomial function and integrates difference with the true solution. The program does so for various mapping orders. 

Situation became even worse in the current development branch when compared to 8.2. 
Here is output for 8.2:

> DEAL::dim: 3   FE_Q<3>(QGaussLobatto(5))
> DEAL::Mapping degree   ||u-u_h||
> DEAL::1    6.8494e-13
> DEAL::2    9.1443e-08
> DEAL::3    1.2932e-06
> DEAL::4    1.3854e-05
> DEAL::5    1.9439e-05

Here is output for dev-branch (version from 18.09.2015):

> DEAL::dim: 3   FE_Q<3>(QGaussLobatto(5))
> DEAL::Mapping degree   ||u-u_h||
> DEAL::1    6.8494e-13
> DEAL::2    0.00028719
> DEAL::3    0.00028103
> DEAL::4    0.00028536
> DEAL::5    0.00028426

This is how error looks like (this is mapping of degree=5):
![error_p 5_manifolds on](https://cloud.githubusercontent.com/assets/8438035/9995882/ae17da16-6084-11e5-8c2f-53fbecbf9471.png)

The problem exists for all types of elements I have tested (FE_Q, FE_RaviartThomas, FE_Nedelec...) and dim=3 (dim=2 gives correct results) as soon as I attach manifold to a cell. Using MappingQ (p, false), i.e. switching off manifolds for inner part of the domain, does not help. Commenting out line

> cell->set_all_manifold_ids(0);

results in correct behaviour:

> DEAL::dim: 3   FE_Q<3>(QGaussLobatto(5))
> DEAL::Mapping degree   ||u-u_h||
> DEAL::1    6.8494e-13
> DEAL::2    6.8495e-13
> DEAL::3    6.8497e-13
> DEAL::4    6.8491e-13
> DEAL::5    6.8492e-13

And this is how error looks like (independent from mapping degree):
![error_p 5_manifolds off](https://cloud.githubusercontent.com/assets/8438035/9995884/b7a0078e-6084-11e5-9aab-7ceca412909f.png)

I do not have much experience with Manifolds, but it seems currently the whole functionality is broken unless one uses MappingQ(1). Can anyone more experienced have a look or advice on how to approach this problem?

Here is the program: https://polybox.ethz.ch/public.php?service=files&t=8ca04791d4f8bd21fce2c54aa2ecad27

**Comment by luca-heltai** [comment]
I'm looking at it right now. 

I've created a new gist with your test:

https://gist.github.com/luca-heltai/9435d4c1a3b4180c6271

**Comment by luca-heltai** [comment]
You are using a ChartManifold with a transformation which is *not* polynomial. Why do you expect integration of polynomials to be exact?

**Comment by luca-heltai** [comment]
Also, you are defining manually a `pull_back` and a `push_forward`. They have to be one the inverse of the other for the `Manifold` object to function properly. Is this the case? I didn't check too deeply, but it looks to me as they are not. In this way, of course also the basic functionalities won't work. 

What is it that you wanted to achieve with this manifold description?

**Comment by agrayver** [comment]
@luca-heltai if by "one the inverse of the other" you mean that condition 
```p == geometry.pull_back(geometry.push_forward(p))```

should hold thoughout domain then one is the inverse of the other because this condition holds. 

> You are using a ChartManifold with a transformation which is not polynomial.

What does that mean? Function zvalue is not a polynomial? Does step-53 have a polynomial transformation? Because the trasnformation I use is essentially a modified version of the code in step-53 with two main changes:
1. The curvature term is removed (for the time being).
2. The transformation is restricted to a single surface inside the domain while the boundaries remain unchanged. Here is what I mean:
![example](https://cloud.githubusercontent.com/assets/8438035/10019276/c381f5f4-613c-11e5-9646-46b8bdd5afc6.png)
Imagine top boundary is sea surface, blue is water and red are sea floor sediments. 
The code I posted is essentially a downscaled version of what you see above. 

If the problem is the transformation (which I still do not understand to be honest), why does it work correctly for MappingQ1 and any Mapping in 2d? And how to make it working in 3d properly?

**Comment by luca-heltai** [comment]
The process is the following: `MappingQ` queries the `Manifold` (which by default is a `FlatManifold`, or, in other words, a bi-linear mapping) where it should put **its own support points** inside the cell/faces/lines/etc. The meat of this process is in the function

`get_intermediate_points_on_object`

https://github.com/dealii/dealii/blob/master/source/fe/mapping_q_generic.cc#L2987

`MappingQ1` does the same, but there the support points are just the vertices (which it already has) so it won't ask anything to anybody. After this you have a polynomial mapping, which is then composed with the basis functions on the reference element to get you the final basis functions.

If the mapping you are using is quasi-affine (as the bi-linear, or Q1, mapping), then the degree of the polynomial of the basis functions is mostly preserved. In general it is not, since the final object you want to integrate is not, in general, polynomial. You are integrating on the reference element something that contains both the polynomials of your basis functions, as well as the determinant of the Jacobian of the mapping, which in general is not a polynomial. Naturally you won't see any of this if the geometry is "flat", i.e., if the support points are aligned. The effect of such jacobian is big when the deformation of the cell is big, but otherwise is well below the errors you show.

I don't know why things should not work in three dimensions, and the errors you show seem to me too big to account for the above phenomenon. Maybe there is an issue in the way internal points are computed in 3d. I'm trying to figure out the difference between the old `Boundary` code and the new `Manifold` functionality...

**Comment by agrayver** [comment]
What you describe may explain type of errors I see with v. 8.2 (left plot in the top message), where one observe some sort of correlation of the error w.r.t. deformations, but with the current development version the errors pattern has changed and error amplitudes became way larger (right plot). Perhaps understanding where these changes come from will can make other things clear?

**Comment by agrayver** [comment]
@luca-heltai I noticed that commit 08d2d30 shows correct output as if the issue had been fixed. I just pulled the master and it seems situation is still the same. Did something broke in between?

**Comment by kronbichler** [comment]
I think this could be related to the bug fix in #2698. @agrayver please check again when that PR is merged (likely tomorrow).

**Comment by agrayver** [comment]
@kronbichler I'm on it.

**Comment by agrayver** [comment]
No, this patch does not seem to help, unfortunately. But the output is slightly different:

deal.II dev (24.06.2016):

> DEAL::dim: 3	FE_Q<3>(4)
> DEAL::Mapping degree	||u-u_h||
> DEAL::1	6.8599e-13
> DEAL::2	0.00057422
> DEAL::3	0.00056202
> DEAL::4	0.00056985
> DEAL::5	0.00056993

deal.II 8.4.1:

> DEAL::dim: 3	FE_Q<3>(QGaussLobatto(5))
> DEAL::Mapping degree	||u-u_h||
> DEAL::1	6.8598e-13
> DEAL::2	0.00057366
> DEAL::3	0.00056101
> DEAL::4	0.00056948
> DEAL::5	0.00056725

Again, dim=2 works fine with both old and new versions.

**Comment by kronbichler** [comment]
Thanks for checking this so quickly. So it seems we need to investigate this more when #2418 gets done.

**Comment by drwells** [comment]
I spent some time looking into this to see if there is any relation to #3734 (there is not). I added some code to the original test program to output rates of convergence.

The summary is that the results are identical and we obtain the right convergence rates, so I believe we have since fixed the outstanding issue in the `MappingQ` code. I attached the output for both branches (8.2 and 8.5pre) below as well as the code.

Here are the results for 8.2:
```
DEAL::FE degree: 1
DEAL::mapping order: 1
[...]
DEAL::Last L2 error: 0.0013019
DEAL::regression slope: 2.0102
DEAL::mapping order: 2
[...]
DEAL::Last L2 error: 0.0013019
DEAL::regression slope: 2.0102
DEAL::mapping order: 3
[...]
DEAL::Last L2 error: 0.0013019
DEAL::regression slope: 2.0102
DEAL::FE degree: 2
DEAL::mapping order: 1
[...]
DEAL::Last L2 error: 1.8144e-05
DEAL::regression slope: 2.9510
DEAL::mapping order: 2
[...]
DEAL::Last L2 error: 1.8144e-05
DEAL::regression slope: 2.9509
DEAL::mapping order: 3
[...]
DEAL::Last L2 error: 1.8144e-05
DEAL::regression slope: 2.9509
DEAL::mapping order: 4
[...]
DEAL::Last L2 error: 1.8144e-05
DEAL::regression slope: 2.9509
DEAL::FE degree: 3
DEAL::mapping order: 1
[...]
DEAL::Last L2 error: 6.2638e-08
DEAL::regression slope: 4.0063
DEAL::mapping order: 2
[...]
DEAL::Last L2 error: 6.2638e-08
DEAL::regression slope: 4.0063
DEAL::mapping order: 3
[...]
DEAL::Last L2 error: 6.2638e-08
DEAL::regression slope: 4.0063
DEAL::mapping order: 4
[...]
DEAL::Last L2 error: 6.2638e-08
DEAL::regression slope: 4.0063
DEAL::mapping order: 5
[...]
DEAL::Last L2 error: 6.2638e-08
DEAL::regression slope: 4.0063
```

and for the development branch:
```
DEAL::FE degree: 1
DEAL::mapping order: 1
[...]
DEAL::Last L2 error: 0.0013019
DEAL::regression slope: 2.0102
DEAL::mapping order: 2
[...]
DEAL::Last L2 error: 0.0013019
DEAL::regression slope: 2.0102
DEAL::mapping order: 3
[...]
DEAL::Last L2 error: 0.0013019
DEAL::regression slope: 2.0102
DEAL::FE degree: 2
DEAL::mapping order: 1
[...]
DEAL::Last L2 error: 1.8144e-05
DEAL::regression slope: 2.9510
DEAL::mapping order: 2
[...]
DEAL::Last L2 error: 1.8144e-05
DEAL::regression slope: 2.9509
DEAL::mapping order: 3
[...]
DEAL::Last L2 error: 1.8144e-05
DEAL::regression slope: 2.9509
DEAL::mapping order: 4
[...]
DEAL::Last L2 error: 1.8144e-05
DEAL::regression slope: 2.9509
DEAL::FE degree: 3
DEAL::mapping order: 1
[...]
DEAL::Last L2 error: 6.2638e-08
DEAL::regression slope: 4.0063
DEAL::mapping order: 2
[...]
DEAL::Last L2 error: 6.2638e-08
DEAL::regression slope: 4.0063
DEAL::mapping order: 3
[...]
DEAL::Last L2 error: 6.2638e-08
DEAL::regression slope: 4.0063
DEAL::mapping order: 4
[...]
DEAL::Last L2 error: 6.2638e-08
DEAL::regression slope: 4.0063
DEAL::mapping order: 5
[...]
DEAL::Last L2 error: 6.2638e-08
DEAL::regression slope: 4.0063
```

Here is the code:
```cpp
#include <deal.II/base/function.h>
#include <deal.II/base/logstream.h>
#include <deal.II/base/quadrature_lib.h>

#include <deal.II/dofs/dof_handler.h>

#include <deal.II/fe/fe_q.h>
#include <deal.II/fe/mapping_q.h>

#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/grid_out.h>
#include <deal.II/grid/tria.h>
#include <deal.II/grid/manifold.h>
#include <deal.II/grid/manifold_lib.h>
#include <deal.II/grid/grid_tools.h>

#include <deal.II/lac/constraint_matrix.h>
#include <deal.II/lac/full_matrix.h>
#include <deal.II/lac/vector.h>

#include <deal.II/numerics/vector_tools.h>
#include <deal.II/numerics/data_out.h>

#include <fstream>
#include <iostream>

using namespace dealii;

// Like FESeries::linear_regression
double regression_slope(const std::vector<double> &x,
                        const std::vector<double> &y)
{
  FullMatrix<double> K(2,2), invK(2,2);
  Vector<double>     X(2), B(2);

  Assert (x.size() == y.size(),
          ExcMessage("x and y are expected to have the same size"));

  Assert (x.size() >= 2,
          dealii::ExcMessage("at least two points are required for linear regression fit"));

  double sum_1 = 0.0,
    sum_x = 0.0,
    sum_x2= 0.0,
    sum_y = 0.0,
    sum_xy= 0.0;

  for (unsigned int i = 0; i < x.size(); i++)
    {
      sum_1  += 1.0;
      sum_x  += x[i];
      sum_x2 += x[i]*x[i];
      sum_y  += y[i];
      sum_xy += x[i]*y[i];
    }

  K(0,0) = sum_1;
  K(0,1) = sum_x;
  K(1,0) = sum_x;
  K(1,1) = sum_x2;

  B(0)   = sum_y;
  B(1)   = sum_xy;

  invK.invert(K);
  invK.vmult(X,B,false);

  return X(1);
}

double zvalue (const double x, const double y)
{
  double xh = x * 5., yh = y * 5.;
  return (xh * exp(-xh*xh - yh*yh)) / 10.;
}

template<int dim>
class Geometry : public ChartManifold<dim>
{
public:
  virtual Point<dim> pull_back(const Point<dim> &space_point) const;
  virtual Point<dim> push_forward(const Point<dim> &chart_point) const;
};

template<int dim>
Point<dim> Geometry<dim>::pull_back(const Point<dim> &space_point) const
{
  const double d = space_point[dim - 1];
  const double z = zvalue(space_point[0], dim == 3 ? space_point[1] : 0);

  double d_hat = 0.;
  if((d - z) <= 0)
    d_hat = (d - z) / (1. + z);
  else
    d_hat = (d - z) / (1. - z);

  Point<dim> p;
  for(unsigned i = 0; i < dim - 1; ++i)
    p[i] = space_point[i];
  p[dim - 1] = d_hat;

  return p;
}

template<int dim>
Point<dim> Geometry<dim>::push_forward(const Point<dim> &chart_point) const
{
  const double d_hat = chart_point[dim - 1];
  const double z = zvalue(chart_point[0], dim == 3 ? chart_point[1] : 0);

  double d = 0.;
  if(d_hat <= 0)
    d = d_hat + (d_hat + 1.) * z;
  else
    d = d_hat - (d_hat - 1.) * z;

  Point<dim> p;
  for(unsigned i = 0; i < dim - 1; ++i)
    p[i] = chart_point[i];
  p[dim - 1] = d;

  return p;
}

// original test case
template<int dim>
class VectorFunction : public Function<dim>
{
public:
  virtual double value (const Point<dim> &p, const unsigned int /*component*/) const
  {
    double result = 1.0;
    for (unsigned int dim_n = 0; dim_n < dim; ++dim_n)
      {
        result *= (1.0 - p[dim_n]*p[dim_n]);
      }
    return result;
  }
};

template<int dim>
class TranscendentalManufacturedSolution : public Function<dim>
{
public:
  virtual double value (const Point<dim> &p, const unsigned int /*component*/) const
  {
    return std::cos(p[0]) + 2.0*std::sin(3*p[1]);
  }
};

template <int dim>
void create_tria(Triangulation<dim>& triangulation, const Geometry<dim>& geometry)
{
  GridGenerator::hyper_cube (triangulation, -1.0, 1.0);
  GridTools::transform (std_cxx11::bind(&Geometry<dim>::push_forward,
                                        std_cxx11::cref(geometry),
                                        std_cxx11::_1),
                        triangulation);

  triangulation.set_manifold(0, geometry);
  for (Triangulation<3>::active_cell_iterator cell=triangulation.begin_active();
       cell!=triangulation.end(); ++cell)
    cell->set_all_manifold_ids(0);

  triangulation.refine_global(3);
}

template <int dim>
void test(const FiniteElement<dim>& fe)
{
  Geometry<dim> geometry;
  TranscendentalManufacturedSolution<dim> fe_function;
  // uncomment the next line and comment the line above to use the biquadratic
  // solution instead of the transcendental one.
  // VectorFunction<dim> fe_function;
  ConstraintMatrix constraints;
  constraints.close();

  deallog << "FE degree: " << fe.degree << std::endl;
  for(unsigned mapping_p = 1; mapping_p < fe.degree + 3; ++mapping_p)
  {
    deallog << "mapping order: " << mapping_p << std::endl;
    Triangulation<dim> triangulation;
    create_tria(triangulation, geometry);
    DoFHandler<dim> dof_handler;

    std::vector<double> log_refinements;
    std::vector<double> log_l2_errors;

    MappingQ<dim> mapping(mapping_p, true);
    for (unsigned int refinement_n = 1; refinement_n < 4; ++refinement_n)
      {
        triangulation.refine_global(1);
        dof_handler.clear();
        dof_handler.initialize(triangulation, fe);

        Vector<double> v(dof_handler.n_dofs());
        VectorTools::project(mapping, dof_handler, constraints,
                             QGauss<dim>(fe.degree + (mapping_p + 2)/2), fe_function, v);

        Vector<double> diff(triangulation.n_active_cells());
        VectorTools::integrate_difference(mapping, dof_handler, v, fe_function, diff,
                                          // superconvergence with QGauss(k + 1)
                                          // QGauss<dim>(fe.degree + 2),
                                          // normal convergence with QIterated
                                          QIterated<dim>(QGauss<1>(fe.degree + 1), 2),
                                          VectorTools::L2_norm);
        log_refinements.push_back(std::log10(std::pow(2.0, -double(refinement_n))));
        log_l2_errors.push_back(std::log10(diff.l2_norm()));
        if (log_refinements.size() > 1)
          {
            deallog << "current slope: "
                    << (*(log_l2_errors.end() - 1) - *(log_l2_errors.end() - 2))
              /(*(log_refinements.end() - 1) - *(log_refinements.end() - 2))
                    << std::endl;
          }
      }
    deallog << "Last number of DoFs: "
            << dof_handler.n_dofs()
            << std::endl;
    deallog << "Last L2 error: "
              << std::pow(10.0, log_l2_errors.back())
              << std::endl;
    deallog << "regression slope: "
              << regression_slope(log_refinements, log_l2_errors)
              << std::endl;
  }
}

int main ()
{
  deallog << std::setprecision (5);
  deallog.attach (std::cout);
  deallog.depth_console (0);

  const static unsigned dim = 3;

  for(unsigned p = 1; p < 4; ++p)
  {
    test<dim>(FE_Q<dim>(QGaussLobatto<1>(p+1)));
  }
}
```

Note that this code uses a transcendental manufactured solution instead of a biquadratic.

The only remaining mystery is that when we use `MappingQ1` with the manifold we do not get exact answers for second order or higher polynomials (when I switch back to the original manufactured solution): this indicates a problem with the order 1 mapping, but the others seem fine. I believe that this should work even when we don't use a manifold description (it works if we turn all of that off).

**Comment by bangerth** [comment]
The outputs you post for 8.2 and dev look the same to me. Was this intentional?

In any case, if you think that the issue is fixed, would you mind making your program into a testcase that we could add to the testsuite, and then close this issue?

**Comment by drwells** [comment]
Yes. I'll run things again to be sure but they had identical output up to resource consumption (@masterleinad @davydden when run with 8.5pre we use 3GB of memory but with 8.2 we use 15GB... nice work with the matrix free stuff!)

**Comment by agrayver** [comment]
@drwells thanks for looking at it. There are two things to mention. First of all, issue that existed in versions > 8.2 was indeed fixed (most likely by #3589  -- but I'm not sure).

The thing that was initially unclear to me is why does error grow with the mapping order for VectorFunction? 

> DEAL::FE degree: 1
> DEAL::mapping order: 1
> DEAL::Last number of DoFs: 35937
> DEAL::Last L2 error: 0.00073770
> DEAL::regression slope: 1.9992
> DEAL::mapping order: 2
> DEAL::Last L2 error: 0.00074526
> DEAL::regression slope: 2.0216
> DEAL::mapping order: 3
> DEAL::Last L2 error: 0.00074458
> DEAL::regression slope: 2.0214
> DEAL::FE degree: 2
> DEAL::mapping order: 1
> DEAL::Last number of DoFs: 274625
> DEAL::Last L2 error: 4.8621e-07
> DEAL::regression slope: 2.2635
> DEAL::mapping order: 2
> DEAL::Last L2 error: 1.2047e-06
> DEAL::regression slope: 2.5491
> DEAL::mapping order: 3
> DEAL::Last L2 error: 1.1048e-05
> DEAL::regression slope: 2.2890
> DEAL::mapping order: 4
> DEAL::Last L2 error: 1.1232e-05
> DEAL::regression slope: 2.2849
> DEAL::FE degree: 3
> DEAL::mapping order: 1
> DEAL::Last number of DoFs: 912673
> DEAL::Last L2 error: 1.6124e-09
> DEAL::regression slope: 2.5439
> DEAL::mapping order: 2
> DEAL::Last L2 error: 5.8920e-08
> DEAL::regression slope: 4.0501
> DEAL::mapping order: 3
> DEAL::Last L2 error: 9.9561e-08
> DEAL::regression slope: 3.6532
> DEAL::mapping order: 4
> DEAL::Last L2 error: 4.6582e-07
> DEAL::regression slope: 4.1645
> DEAL::mapping order: 5
> DEAL::Last L2 error: 4.5760e-07
> DEAL::regression slope: 4.1834

Now I see that with your TranscendentalManufacturedSolution the error remains constant and the slope looks correct. So I tend to that maybe the situation has to do with the particular function that we integrate and not deal.II.

**Comment by drwells** [comment]
@agrayver I am not sure why this error increases. My best guess is that there is some strange effect since the exact solution is biquadratic: without the mapping we would be able to represent it exactly. I am going to run a few more tests to try and understand this better (mainly, at this point, because it is interesting and I want to know).

**Comment by drwells** [comment]
@agrayver I spent a lot of time over the last week reading about superparametric finite elements and I have not yet come up with a good reason for why the error grows slightly as the mapping order increases. My best explanation is this: if we hold the order of the FE constant and increase the order of the mapping then, in a sense, we have made the problem harder since we are trying to approximate something more complex (harder geometry) with what is fundamentally the same FE space. This is the opposite of the usual condition (where, perhaps, we would solve things on a disk where the solution is known to be zero on the boundary) since the choice of boundary here (with a manufactured solution) is arbitrary.

Regardless, things seem to work now, so I am closing this (any related problems will surely be discussed in #2473).
